File: C++/pilo.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

// TODO: Please Add headers definitions to improve code organization

namespace Pilo {
    bool hadError = false;

    void runFile(const std::string& path);
    void runPrompt();
    void error(int line, const std::string& message);
    void report(int line, const std::string& where, const std::string& message);
    void run(const std::string& source);

    void runFile(const std::string& path) {
        std::ifstream file(path);
        std::stringstream buffer;
        buffer << file.rdbuf();
        run(buffer.str());
        if (hadError) {
            exit(65);
        }
    }

    void runPrompt() {
        std::string line;
        for (;;) {
            std::cout << "> ";
            if (!std::getline(std::cin, line)) {
                break;
            }
            run(line);
            hadError = false;
        }
    }

    void error(int line, const std::string& message) {
        report(line, "", message);
    }

    void report(int line, const std::string& where, const std::string& message) {
        std::cerr << "[line " << line << "] Error" << where << ": " << message << std::endl;
        hadError = true;
    }

    void run(const std::string& source) {
        // TODO: Implement the Pilo interpreter logic
    }
}


int main(int argc, char** argv) {
    if (argc > 2) {
        std::cout << "Pilo language: " << std::endl;
        return 64;
    } else if (argc == 2) {
        Pilo::runFile(argv[1]);
    } else {
        Pilo::runPrompt();
    }

    return 0;
}

File: JavaScript/pilo.js
const fs = require('fs');
const readline = require('readline');

let hadError = false;


//TODO: Put all these in a namespace or let's use typescript maybe?
function runFile(path) {
  const bytes = fs.readFileSync(path);
  run(bytes.toString());
  if (hadError) process.exit(65);
}

function runPrompt() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  rl.on('line', (line) => {
    run(line);
    hadError = false;
    console.log('> ');
  });
}

function error(line, message) {
  report(line, '', message);
}

function report(line, where, message) {
  console.error(`[line ${line}] Error${where}: ${message}`);
  hadError = true;
}

function run(source) {
  //TODO: run logic
}

const args = process.argv.slice(2);

if (args.length > 1) {
  console.log('Pilo language:');
  process.exit(64);
} else if (args.length === 1) {
  runFile(args[0]);
} else {
  runPrompt();
}